stl/inc/syncstream

```diff
@@ -0,0 +1,339 @@
+        mutex _Mutex;
+        uint64_t _Ref_count;
+
+        _Mutex_count_pair() : _Ref_count(0), _Mutex(mutex{}) {}
+        ~_Mutex_count_pair()                        = default;
+        _Mutex_count_pair(_Mutex_count_pair&&)      = delete;
+        _Mutex_count_pair(const _Mutex_count_pair&) = delete;
+        _Mutex_count_pair& operator=(const _Mutex_count_pair&) = delete;
+        _Mutex_count_pair& operator=(_Mutex_count_pair&&) = delete;
+    };
+    static map<void*, _Mutex_count_pair>& _Get_mutex_map() noexcept {
+        static map<void*, _Mutex_count_pair> _Mutex_Map;
+        return _Mutex_Map;
+    }
+    static mutex& _Get_map_mutex() noexcept {
+        static mutex _Mutex;
+        return _Mutex;
+    }
+    static mutex& _Get_mutex_for_instance(void* _Ptr) {
+        mutex& _Mutex = _Get_map_mutex();
+        {
+            lock_guard<mutex> _Guard(_Mutex);
+
+            auto& _Mutex_map         = _Get_mutex_map();
+            auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+
+            _STL_ASSERT(_Instance_mutex_itr != _Mutex_map.end(), "No mutex exists for given instance!");
+            return _Instance_mutex_itr->second._Mutex;
+        }
+    }
+    static void _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept {
+        mutex& _Mutex = _Get_map_mutex();
+        {
+            lock_guard<mutex> _Guard(_Mutex);
+
+            auto& _Mutex_map = _Get_mutex_map();
+            _Mutex_map.emplace(std::piecewise_construct, std::forward_as_tuple(_Ptr), std::forward_as_tuple())
+                .first->second._Ref_count++;
+        }
+    }
+    static void _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept {
+        mutex& _Mutex = _Get_map_mutex();
+        {
+            lock_guard<mutex> _Guard(_Mutex);
+
+            auto& _Mutex_map         = _Get_mutex_map();
+            auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+            if (_Instance_mutex_itr == _Mutex_map.end())
+                return;
+            _Instance_mutex_itr->second._Ref_count--;
+            if (_Instance_mutex_itr->second._Ref_count == 0) {
+                _Mutex_map.erase(_Ptr);
+            }
+        }
+    }
+
+    template <class _Elem, class _Traits = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
+    friend class basic_syncbuf;
+};
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits = char_traits<_Elem>, class _Alloc = allocator<_Elem>>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() : basic_syncbuf(nullptr){};

```

Comment:
@miscco, you were saying that this could be `default`. I understand that it is required to initialize the allocator using `_Alloc{}` for a default constructed `basic_syncbuf`.
Can this be achieved using member initialization?
```C++
basic_syncbuf() = default;
// ...
streambuf_type* _Wrapped = nullptr;
allocator_type _Al = _Alloc{};
bool _Emit_on_sync = false;
bool _Record_sync  = false;
```

Reply:
You can just say `allocator_type _Al{};` as either brace-initialization or copy-initialization (with `=`) are acceptable syntax for data member initializers.